### PR TITLE
Issue#314 Replace Logback with Log4j2

### PR DESCRIPTION
**Pull request by badalsarkar**
Implements #314 .

Logback is replaced with Log4j2. Log4j2 has been configured to do logging in async manner. 
I have used system property to dynamically set the log file name [here](https://github.com/paypal/butterfly/compare/develop...badalsarkar:issue%23314?expand=1#diff-75bb75a8f98736940b7246e3aa7999f717f54d8c5aa7e61e5db30d2068452ba5R92).

I have also configured the loggers to be synchronous for testing purpose. 

Please let me know if any change is needed. 

Thanks,

### CODE REVIEW

**Comment by fabiocarvalho777** [comment]
@badalsarkar thank you so much for your contribution. I am on vacation this week, so it might take me a few days to review your PR, but I will add a reminder in my calendar. Thanks again.

By the way, is there any other issue that sounds interesting to you to work with?

**Comment by badalsarkar** [comment]
> @badalsarkar thank you so much for your contribution. I am on vacation this week, so it might take me a few days to review your PR, but I will add a reminder in my calendar. Thanks again.
> 
> By the way, is there any other issue that sounds interesting to you to work with?

Thanks Fábio. It was great to be able to contribute. :smiley: 

I was actually going through the issues to pick one to work next. There are many that are interesting to me. I will let you know soon.

File: butterfly-cli/src/test/java/com/paypal/butterfly/cli/logging/LoggerTest.java
Comment (fabiocarvalho777) [inline butterfly-cli/src/test/java/com/paypal/butterfly/cli/logging/LoggerTest.java:31]: Let's use `assertEquals` here please

File: butterfly-test/build.gradle
Comment (fabiocarvalho777) [inline butterfly-test/build.gradle:10]: Please fix indentation

**Review by fabiocarvalho777 — CHANGES_REQUESTED** [review]

**Comment by fabiocarvalho777** [comment]
@badalsarkar finally reviewed your PR. Added some comments, they are all small. Please address them, or let me know if you have any question. Thanks again for your contribution!

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/logging/LogFileDefiner.java
Comment (badalsarkar) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/logging/LogFileDefiner.java:48]: Yes. I actually removed this file altogether but don't know why git is showing it as modified. Just a question- this file has `getLogFile()` method which is called from [these](https://github.com/paypal/butterfly/search?q=getLogFile) places, `setLogFileName()` which is called from [these](https://github.com/paypal/butterfly/search?q=setLogFileName) places and `setButterflyHome()` which is called from [these](https://github.com/paypal/butterfly/search?q=setButterflyHome) places. If I delete this file, I need to modify these calls. Is this the desired approach? I was not sure, so I did not delete these methods.

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java
Comment (badalsarkar) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: This is absolutely not the right place to for this line of code. I realise now. 
In case of Logback, the file name was retrieved when the framework initialises by calling [`getPropertyValue()`](https://github.com/paypal/butterfly/blob/d7c9819de9e7516d32950b2348c71b9d424d691a/butterfly-cli/src/main/java/com/paypal/butterfly/cli/logging/LogFileDefiner.java#L51). 

In Log4j2 I could not find a way to set the log file name during initialisation by calling this function. So, I used system property. The system property value should be update before any calls to Logger. So, I think, this method should be called [here](https://github.com/paypal/butterfly/blob/d7c9819de9e7516d32950b2348c71b9d424d691a/butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java#L46) right after we set the `butterflyHome`. Later on, when we update the log file name, we can updated the system property [here](https://github.com/paypal/butterfly/blob/d7c9819de9e7516d32950b2348c71b9d424d691a/butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java#L84) so that new file name is used by logger. What do you think? This may also depend on what we decide to do with the `LogFileDefiner` class which you mentioned below.

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/logging/LogFileDefiner.java
Comment (fabiocarvalho777) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/logging/LogFileDefiner.java:48]: I see, I didn't notice removing this file would require all those other changes. Well, in this case it seems your original solution works better. I guess what we need to do is just assume we have repurposed this file now. And I like the fact that we can keep the contract with all those other lines of code that call it. Let's leave it like this then. Thanks for clarifying it.

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java
Comment (fabiocarvalho777) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: Ok, we got a conclusion about the situation with `LogFileDefiner` (see comment below). Now let's go ahead and address this too. Also, is setting the System property really the only way to set the log file? Isn't there a way to do it using Log4j 2 API?

Comment (badalsarkar) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: Let me do some work on that. I will get back soon.

**Comment by fabiocarvalho777** [comment]
@badalsarkar answered your comments.

When you provide your next commit, please don't squash it, this way the code review process keeps the history of the changes as we discuss and they get addressed. GitHub will squash everything into a single commit when I merge the PR.
Also, please don't mark comments as resolved. That is up to the reviewer to do. This helps to keep track of what is pending what is not. If you want, you can add a comment like "done" once you addressed it. Thanks :-)

**Comment by badalsarkar** [comment]
> @badalsarkar answered your comments.
> 
> When you provide your next commit, please don't squash it, this way the code review process keeps the history of the changes as we discuss and they get addressed. GitHub will squash everything into a single commit when I merge the PR.
> Also, please don't mark comments as resolved. That is up to the reviewer to do. This helps to keep track of what is pending what is not. If you want, you can add a comment like "done" once you addressed it. Thanks :-)

Oh! sorry about that. I am new in open source development. Still learning. I will keep that in mind and also the squashing thing. Thanks. :-)

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java
Comment (badalsarkar) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: I looked more into the documentations. Here is what I found so far.

Log4j2 doesn't have any API to directly set the value of `filename` dynamically. I checked [FileAppender](https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/appender/FileAppender.html) and [RandomAccessFileAppender](https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/appender/RandomAccessFileAppender.html). So, we have two options- use property substitution or programmatically configure Log4j2. 

There are different property substitution. Please read [here](https://logging.apache.org/log4j/2.x/manual/configuration.html#PropertySubstitution) and [here](https://logging.apache.org/log4j/2.x/manual/lookups.html). I have used `system` lookup. If we don't want to use `System` property, we can create a [lookup plugin](https://logging.apache.org/log4j/2.x/manual/plugins.html#Lookups) and use it to get the `logFile` name dynamically. The difference between this approach and using `system` property is that with `system` property, we need to set it but with custom lookup plugin, Log4j2 will load the value during initialization. However, to implement it we need to define a class which implements [`StrLookup`](https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/lookup/StrLookup.html) interface. Probably, we can user our `LogFileDefiner` class to implement this interface and implement two abstract methods. It is similar to what was done for Logback. 

If we want to configure programmatically, we can initialize Log4j2 [combining xml configuration and programmatic configuration](https://logging.apache.org/log4j/2.x/manual/customconfig.html#Initialize_Log4j_by_Combining_Configuration_File_with_Programmatic_Configuration). Or [remove the appender and add a new one to root logger](https://logging.apache.org/log4j/2.x/manual/customconfig.html#Programmatically_Modifying_the_Current_Configuration_after_Initialization) which doesn't seem right to me.

To me using a custom lookup plugin seem a right approach. I think I have not missed anything. But yes, let me know if there are other approaches which I am not aware of. I remember, I spent most time to figure this out.

Comment (fabiocarvalho777) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: Hey. Is this you? https://www.linkedin.com/in/badalsarkar/

Comment (badalsarkar) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: Yes. That's me. :-)

Comment (badalsarkar) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: Lets connect in LinkedIn.

Comment (fabiocarvalho777) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: Done ;-)

Comment (badalsarkar) [reply butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliApp.java:92]: Implementation done and is working as expected. I will soon push it.

File: butterfly-cli/src/test/resources/log4j2-test.xml
Comment (badalsarkar) [inline butterfly-cli/src/test/resources/log4j2-test.xml:24]: I need some guidance from you here. This configuration file is used for test purpose. The difference between this and the [production configuration](https://github.com/paypal/butterfly/pull/367/files#diff-3b2b25e9fb51ff13c63c97f5cfde8824e921ad41f2cb52f6bb7edc39ee681b2a) file is that loggers are synchronous here. I did this because the production loggers are async, many test cases were failing. But there is something that concerns me.

- Now there are two configuration files. One for test and one for production. There is possibility they might diverse. For example, if someone change the test file but doesn't make the same change in the production file. To guard against this I have used [this test](https://github.com/paypal/butterfly/pull/367/files#diff-1e04d91dccfdf1522d572997ab5dadfe1d8dc8c82c6c68b81e6e4fda550e4033) which uses the production configuration file. Do you think another approach would have been better? How can we test async loggers?

- Log4j2 has an option called [`XInclude`](https://logging.apache.org/log4j/2.x/manual/configuration.html#XInclude) which allows to add fragments of configuration from different file. This can improve the situation. For example, I can keep the common configuration for production and test in one file and then include the configuration that is varying, separately. The production config is in `butterfly-cli/src/main/resources` folder. Whereas test configuration is in `butterfly/cli/src/test/resources` folder. I am not able to use `XInclude` to include configuration from production folder because relative referencing is not working in `href`. Do you know how to do it?

Comment (fabiocarvalho777) [reply butterfly-cli/src/test/resources/log4j2-test.xml:24]: I see your point, thanks for sharing it. Indeed there is a risk we introduce inconsistencies as we maintain these log files. However, after this PR is merged, unless if we find any bug with this solution, there will be very rare moments when we have to go back here. Also, trying to solve this problem might be too complicated, and not worth it compared to the value it would provide. Having said that, I think it is ok to have this as it is now. So, I think we will be fine.

Still, just to help out a little. Please add a comment to the top of both files explaining the difference between them, why we need a test one, and reminding code maintainers to make sure they are "in sync" whenever any change is necessary.

Thank you very much.

Comment (badalsarkar) [reply butterfly-cli/src/test/resources/log4j2-test.xml:24]: Great. Will do. Thank you. :-)

**Review by fabiocarvalho777 — APPROVED** [review]

**Pull request merged**
